crates/c/src/lib.rs

```diff
@@ -36,6 +37,47 @@ pub extern "C" fn qibo_core_circuit_n_elements(circuit: &Circuit) -> usize {
     circuit.n_elements()
 }
 
+#[no_mangle]
+pub extern "C" fn qibo_core_circuit_n_gates(circuit: &Circuit) -> usize {
+    circuit.n_gates()
+}
+
+#[no_mangle]
+pub extern "C" fn qibo_core_circuit_gate(circuit: &Circuit, gid: usize) -> *const c_char {
+    let gate = circuit.gate(gid);
+    let name = match gate {
+        Gate::One(One::H) => "H",
+        Gate::One(One::X) => "X",
+        Gate::One(One::Y) => "Y",
+        Gate::One(One::Z) => "Z",
+        Gate::Two(Two::CNOT) => "CNOT",
+        _ => todo!()
+    };
+
+    CString::new(name).unwrap().into_raw()
+}
+
+#[no_mangle]
+pub extern "C" fn qibo_core_circuit_elements(circuit: &Circuit, gid: usize, ptr: *mut *const usize, len: *mut usize) {
+    let elements = circuit.elements(gid);
+    // Complaints about what follows are to be directed to ChatGPT
+    let boxed_slice = elements.clone().into_boxed_slice();
+    unsafe {
+        *ptr = boxed_slice.as_ptr();
+        *len = elements.len();
+    }
+    std::mem::forget(boxed_slice);
+}
+
+#[no_mangle]
+pub extern "C" fn qibo_core_circuit_free_elements(ptr: *const usize, len: usize) {
+    if !ptr.is_null() {
+        unsafe {
+            let _ = Vec::from_raw_parts(ptr as *mut usize, len, len);
+        }
+    }
+}
```

Comment:
This function should not be needed at all. You can leave the memory management of Rust objects to the Rust library (creators and destructors are defined, as in C++ standard library, such that you should not allocate memory explicitly, following RAII philosophy).

Reply:
That's wrong: since the program is written in C, the memory handling is up to C itself.

In C++, we could have used the destructors, and default ones provided by the STL. In C, there is nothing like that, thus we have to manually free the memory we allocate.
We will need to do this even for circuits, strings, and everything else.

Reply:
Cf. #38